### PR TITLE
Detach and Terminate Ec2

### PR DESCRIPTION
**Pull request by Joeper214**
Ref: https://skyarch.backlog.jp/view/NC-496

### Goal for this PR:

- [x] Detach Ec2 from skyhopper service.
- [x] Terminate Ec2.

### CODE REVIEW

File: app/controllers/ec2_instances_controller.rb [hunk 1]
```diff
@@ -77,6 +77,39 @@ def stop
     render text: I18n.t('ec2_instances.msg.stop_ec2')
   end
 
+  # POST /ec2_instances/i-hogehoge/detach
+  def detach
+    physical_id = params.require(:id)
+    infra_id    = params.require(:infra_id)
```
Comment (skyarch-networks) [inline app/controllers/ec2_instances_controller.rb]: __[rubocop]__
* Lint/UselessAssignment: Useless assignment to variable - `infra_id`. [:ref](https://github.com/bbatsov/ruby-style-guide#underscore-unused-vars)

File: app/controllers/ec2_instances_controller.rb [hunk 2]
```diff
@@ -77,6 +77,38 @@ def stop
...
+    resource.destroy
+
+    infra_logger_success("#{physical_id} has been detached.")
+
+    notify_ec2_status(resource, :detached)
+
+    render text: I18n.t('ec2_instances.msg.detach_ec2')
+  end
+
+  # POST /ec2_instances/i-hogehoge/detach
```
Comment (pocke) [inline app/controllers/ec2_instances_controller.rb]: Please rename "detach" to "terminate" in comment.

Comment (Joeper214) [inline app/controllers/ec2_instances_controller.rb]: :+1:

**Comment by pocke** [comment]
I think EC2 should be detached from zabbix and chef server in `detach` and `terminate` actions.
For reference, Infra is detached from those services in `infrastructure#destroy` action.
What do you think?

**Comment by Joeper214** [comment]
Yeah. I think so. Sorry I did forgot to examine those cases. Thank you for the suggestion. I'll implement this.

**Comment by Joeper214** [comment]
Test failed. :scream_cat:

**Comment by pocke** [comment]
I'll try to write test

**Comment by Joeper214** [comment]
Thanks.

**Comment by Joeper214** [comment]
Wait there's still an error for this..


chef_id

**Comment by pocke** [comment]
Please merge develop and resolve conflict :bow:

File: app/models/node.rb [hunk 3]
```diff
@@ -287,4 +292,22 @@ def exec_knife_ssh(command, infra)
   ensure
     ec2key.close_temp
   end
+
+  def exec_knife_ssh_noblock(command, infra)
+    ec2key = infra.ec2_private_key
+    puts ec2key.output_temp(prefix: @name)
+    fqdn = infra.instance(@name).fqdn
```
Comment (skyarch-networks) [inline app/models/node.rb]: __[brakeman]__
* Possible command injection [:ref](http://brakemanscanner.org/docs/warning_types/command_injection/)

File: app/models/node.rb [hunk 4]
```diff
@@ -271,7 +276,7 @@ def exec_knife_ssh(command, infra)
     fqdn = infra.instance(@name).fqdn
 
     cmd = "ssh #{@user}@#{fqdn} -t -t -i #{ec2key.path_temp} #{command}"
-
+    puts cmd
```
Comment (skyarch-networks) [inline app/models/node.rb]: __[rubocop]__
* Rails/Output: Do not write to stdout. Use Rails' logger if you want to log.

File: app/models/node.rb [hunk 5]
```diff
@@ -287,4 +292,22 @@ def exec_knife_ssh(command, infra)
   ensure
     ec2key.close_temp
   end
+
+  def exec_knife_ssh_noblock(command, infra)
+    ec2key = infra.ec2_private_key
+    puts ec2key.output_temp(prefix: @name)
```
Comment (skyarch-networks) [inline app/models/node.rb]: __[rubocop]__
* Rails/Output: Do not write to stdout. Use Rails' logger if you want to log.

File: app/models/node.rb [hunk 6]
```diff
@@ -287,4 +292,22 @@ def exec_knife_ssh(command, infra)
...
+    fqdn = infra.instance(@name).fqdn
+
+    cmd = <<-EOS
+            ssh #{@user}@#{fqdn} -t -t -i #{ec2key.path_temp} #{command}
+          EOS
+    puts cmd
```
Comment (skyarch-networks) [inline app/models/node.rb]: __[rubocop]__
* Rails/Output: Do not write to stdout. Use Rails' logger if you want to log.

File: lib/tasks/test.rake [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# lib/tasks/test.rake
+namespace :crono do
+  desc 'Update all tables'
+  task :hello => :environment do
```
Comment (skyarch-networks) [inline lib/tasks/test.rake]: __[rubocop]__
* Style/HashSyntax: Use the new Ruby 1.9 hash syntax. [:ref](https://github.com/bbatsov/ruby-style-guide#hash-literals)

File: app/models/node.rb [hunk 8]
```diff
@@ -79,12 +79,25 @@ def delete
     ChefAPI.destroy('client', @name)
   end
 
+  def delete_node
+    cmd = <<-EOS
+            knife node delete #{@name} -y
+    EOS
+    puts cmd
```
Comment (skyarch-networks) [inline app/models/node.rb]: __[rubocop]__
* Rails/Output: Do not write to stdout. Use Rails' logger if you want to log.

**Comment by Joeper214** [comment]
Since, @pocke is not here today.
 

I will assign review and merging of this task to @root-beer -san. Thanks!  :bow:

**Comment by Joeper214** [comment]
## TODO:

- [ ] Reuse Modal form confirmation of action
- [ ] Show UI for Optional  detaching Zabbix and Chef Server as checbox
- [ ] iT18N

**Comment by Joeper214** [comment]
## TODO:

- [ ] Reuse Modal form confirmation of action

- [ ] iT18N 

## Next goal
- [ ] Show UI for Optional  detaching Zabbix and Chef Server as checbox
- [ ] Back-end filtering if Zabbix or chef is to be detached (According to input).

**Comment by Joeper214** [comment]
Todo done. Please review and Merge PR: @root-beer . Thanks. :bow:

**Comment by root-beer** [comment]
I review it from now.

**Comment by Joeper214** [comment]
Thanks.

File: app/models/resource.rb [hunk 9]
```diff
@@ -45,4 +48,20 @@ def initialize_statuses
...
+    s = AppSetting.get
+    begin
+      z = Zabbix.new(s.zabbix_user, s.zabbix_pass)
+      z.delete_hosts_by_resource(self.physical_id)
+    rescue => ex
+      return self if ex.message.include("physical id not found.")
+      raise ex
+    end
+  end
+  self
```
Comment (root-beer) [inline app/models/resource.rb]: What does this `self` mean?

Comment (Joeper214) [inline app/models/resource.rb]: hmm.. sorry It shouldn't be there.

**Comment by root-beer** [comment]
LGTM.

**Comment by root-beer** [comment]
Merged it. Thanks!

**Comment by Joeper214** [comment]
Thanks!

**Pull request merged**